shims/spark34/src/main/scala/org/apache/spark/TaskContextUtils.scala

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark
+
+import org.apache.spark.executor.TaskMetrics
+import org.apache.spark.memory.{TaskMemoryManager, UnifiedMemoryManager}
+import org.apache.spark.metrics.MetricsSystem
+import org.apache.spark.network.util.ByteUnit
+
+import java.util.Properties
+
+object TaskContextUtils {
```

Comment:
Why is it necessary to place this class in the shim layer?

Reply:
It's needed mainly for 2 reasons:

1. Spark requires `SparkEnv` to be set to instantiate `TaskContextImpl` in its later version. Since we don't plan to set `SparkEnv` which is not thread-local in `runUnsafe` routine, we should manually pass `cpus` to skip `SparkEnv.get`
https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/TaskContextImpl.scala#L58

2. Spark 3.4 added `numPartitions` to the arg list
https://github.com/apache/spark/commit/f48602912b4d967a59085226345e056bc39d865e